src/tokens/ERC1155.vy

```diff
@@ -178,57 +270,207 @@ def balanceOfBatch(owners: DynArray[address, max_value(uint16)], ids: DynArray[u
     """
     @dev Batched version of `balanceOf`.
     @notice Note that `owners` and `ids` must have the
-            same length.
+        same length.
     @param owners The 20-byte array of owner addresses.
     @param ids The 32-byte array of token identifiers.
     @return DynArray The 32-byte array of token amounts
-            owned by `owners`.
+        owned by `owners`.
     """
-    return empty(DynArray[uint256, max_value(uint16)])
+    assert len(owners) == len(ids), "ERC1155: batch lengths mismatch"
+    batch_balances: DynArray[uint256, max_value(uint16)] = []
+    for i in range(max_value(uint16)):
+        # if the owner address is zero address, we break the loop to save gas
+        if owners[i] == empty(address): break
```

Comment:
would not be needed if we directly call `balanceOf` above and include the zero address check there.

Reply:
What do you suggest, that `balanceOf` is a view function instead of a public variable?

The zero address may never have a non-zero balance, so we could just let it iterate over it and load zero from storage.

Reply:
right, it would look like this (not reflected our own indexing strategy described above):

```vyper
batch_balances[i] = IERC1155(self).balanceOf(owners[i], ids[i])
```

thoughts?

Reply:
Sounds good, updating